src/Agda2Hs/Compile/RuntimeCheckUtils.hs

```diff
@@ -0,0 +1,252 @@
+createRtc :: Hs.Name () -> [Hs.Pat ()] -> [Hs.Exp ()] -> Hs.Exp () -> Maybe (Hs.Binds ()) -> Hs.Decl ()
+createRtc n args [] success = createRtc' n args $ Hs.UnGuardedRhs () success
+createRtc n args chks success =
+  createRtc' n args rhs
+  where
+    rhs = Hs.GuardedRhss () $ Hs.GuardedRhs () [Hs.Qualifier () andChks] success : notChks ++ [otherwiseChk]
+    andChks = foldr1 (\c -> Hs.InfixApp () c $ Hs.QVarOp () $ Hs.UnQual () $ Hs.Symbol () "&&") chks
+    errs =
+      zip chks $
+        map
+          ( \chk ->
+              let msg = "Runtime check failed: " ++ Hs.prettyPrint chk
+               in Hs.App () (hsVar "error") $ Hs.Lit () $ Hs.String () msg msg
+          )
+          chks
+    (nots, otherwise) = fromJust $ initLast errs
+    notChks = map (\(chk, err) -> Hs.GuardedRhs () [Hs.Qualifier () $ Hs.App () (hsVar "not") chk] err) nots
+    otherwiseChk = (\(_, err) -> Hs.GuardedRhs () [Hs.Qualifier () $ hsVar "otherwise"] err) otherwise
+
+createRtc' :: Hs.Name () -> [Hs.Pat ()] -> Hs.Rhs () -> Maybe (Hs.Binds ()) -> Hs.Decl ()
+createRtc' n args rhs binds = Hs.FunBind () [Hs.Match () n args rhs binds]
+
+type NameIndices = (Int, Int)
+
+-- Creates a runtime check if necessary and possible, informing C accordingly.
+-- Takes telescope of type to check, name, and expression on success.
+checkRtc :: Telescope -> QName -> Hs.Exp () -> C RtcResult
+checkRtc tel name success =
+  checkRtc' (0, 0) tel
+    >>= \case
+      NoneErased' -> return NoneErased
+      Uncheckable' -> return Uncheckable
+      Checkable' {..} -> do
+        tellAllCheckable name
+        let rhs = eApp success theirRhs
+            chkName = hsName $ prettyShow $ qnameName name
+            chk = createRtc chkName theirLhs theirChks rhs $ binds theirDecls
+        return $ Checkable [chk]
+      . snd
+
+-- Recursively check for runtime checkability in nested types.
+-- Accumulates on name indices for `go` function and `a` argument.
+-- Takes telescope of type to check.
+checkRtc' :: NameIndices -> Telescope -> C (NameIndices, RtcResult')
+checkRtc' idcs tel = do
+  ourChks <- mapM (uncurry createGuardExp) topLevelErased
+  (belowIdcs, belowChks) <- mapAccumLM checkRtc'' idcs call
+  (belowIdcs,)
+    <$> if all isJust belowChks && all isJust ourChks
+      then
+        let (theirLhs, theirRhs, decls) = unzip3 $ catMaybes belowChks
+            theirDecls = concat decls
+            theirChks = catMaybes ourChks
+         in if null theirDecls && null ourChks
+              then return NoneErased'
+              else return Checkable' {..}
+      else return Uncheckable'
+  where
+    doms = telToList tel
+    telsUpTo = map (\i -> fst $ splitTelescopeAt i tel) [0 ..]
+    -- Partition out arguments that are erased and at top level (those we will attempt to check)
+    (topLevelErased, call) = partition (checkTopLevelErased . fst) $ zip doms telsUpTo
+
+-- Check a single type for runtime checkability.
```

Comment:
Since this is the core logic of this PR, I would appreciate some more detailed commentary with possibly an example to make clear what the `NameIndices` are for. Adding line-by-line explanations to the function definitions would also be a great help, since they are quite difficult to grasp at the moment.